org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JdtSourceLookUpProvider.java

```diff
@@ -0,0 +1,139 @@
+     */
+    private Object findSourceElement(String sourcePath, ISourceContainer[] containers) {
+        if (containers == null) {
+            return null;
+        }
+        for (ISourceContainer container : containers) {
+            try {
+                Object[] objects = container.findSourceElements(sourcePath);
+                if (objects.length > 0 && (objects[0] instanceof IResource || objects[0] instanceof IClassFile)) {
+                    return objects[0];
+                }
+            } catch (CoreException e) {
+                JavaLanguageServerPlugin.logException("Failed to find the source elements", e);
+            }
+        }
+        return null;
+    }
+
+    private ISourceContainer[] getSourceContainers(List<String> projectNames) {
+        List<IProject> projects = new ArrayList<>();
+        if (projectNames == null || projectNames.size() == 0) {
+            projects.addAll(Arrays.asList(ProjectUtils.getAllProjects()));
+        } else {
+            for (String projectName : projectNames) {
+                projects.add(ProjectUtils.getProject(projectName));
+            }
+        }
+
+        Set<ISourceContainer> containers = new LinkedHashSet<>();
+        Set<IRuntimeClasspathEntry> calculated = new LinkedHashSet<>();
+        projects.stream().map(project -> ProjectUtils.getJavaProject(project))
+            .filter(javaProject -> javaProject != null && javaProject.exists())
+            .forEach(javaProject -> {
+                // Add source containers associated with the project's runtime classpath entries.
+                containers.addAll(Arrays.asList(getSourceContainers(javaProject, calculated)));
+                // Add source containers associated with the project's source folders.
+                containers.add(new JavaProjectSourceContainer(javaProject));
+            });
+
+        return containers.toArray(new ISourceContainer[0]);
+    }
+
+    private ISourceContainer[] getSourceContainers(IJavaProject project, Set<IRuntimeClasspathEntry> calculated) {
+        if (project == null || !project.exists()) {
+            return new ISourceContainer[0];
+        }
+
+        try {
+            IRuntimeClasspathEntry[] unresolved = JavaRuntime.computeUnresolvedRuntimeClasspath(project);
+            List<IRuntimeClasspathEntry> resolved = new ArrayList<>();
+            for (IRuntimeClasspathEntry entry : unresolved) {
+                for (IRuntimeClasspathEntry resolvedEntry : JavaRuntime.resolveRuntimeClasspathEntry(entry, project)) {
+                    if (!calculated.contains(resolvedEntry)) {
+                        calculated.add(resolvedEntry);
+                        resolved.add(resolvedEntry);
+                    }
+                }
+            }
+            Set<ISourceContainer> containers = new LinkedHashSet<>();
+            containers.addAll(Arrays.asList(
+                    JavaRuntime.getSourceContainers(resolved.toArray(new IRuntimeClasspathEntry[0]))));
```

Comment:
I think there may be an issue here when JavaSourceLookupUtil#translate(..) creates a PackageFragmentRootSourceContainer(..) for the archive entries. The PackageFragmentRootSourceContainer class doesn't seem to use the source attachment to figure the source location. It ends up just returning a reference in the classfile.

Reply:
Yes, here we mean to resolve a uri for a given stack trace line, and the uri comes from the classfile's reference.

To resolve the content(source code) of the uri, we have a content provider to do this: https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/DisassemblerContentProvider.java#L52